File: qc-social-api/src/main/java/com/eager/questioncloud/social/SocialAPIManager.java
package com.eager.questioncloud.social;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class SocialAPIManager {
    private final List<SocialAPI> socialAPIs;

    public String getAccessToken(String code, SocialPlatform socialPlatform) {
        for (SocialAPI socialAPI : socialAPIs) {
            if (socialAPI.getSocialPlatform().equals(socialPlatform)) {
                return socialAPI.getAccessToken(code);
            }
        }
        throw new FailSocialLoginException();
    }

    public String getSocialUid(String accessToken, SocialPlatform socialPlatform) {
        for (SocialAPI socialAPI : socialAPIs) {
            if (socialAPI.getSocialPlatform().equals(socialPlatform)) {
                return socialAPI.getUserInfo(accessToken).uid();
            }
        }
        throw new FailSocialLoginException();
    }
}

File: qc-social-api/src/main/java/com/eager/questioncloud/social/SocialAPIManager.kt
package com.eager.questioncloud.social

import org.springframework.stereotype.Component

@Component
class SocialAPIManager(
    private val socialAPIs: List<SocialAPI>
) {
    fun getAccessToken(code: String, socialPlatform: SocialPlatform): String {
        for (socialAPI in socialAPIs) {
            if (socialAPI.getSocialPlatform() == socialPlatform) {
                return socialAPI.getAccessToken(code)
            }
        }
        throw FailSocialLoginException()
    }

    fun getSocialUid(accessToken: String, socialPlatform: SocialPlatform): String {
        for (socialAPI in socialAPIs) {
            if (socialAPI.getSocialPlatform() == socialPlatform) {
                return socialAPI.getUserInfo(accessToken).uid
            }
        }
        throw FailSocialLoginException()
    }
}
